casanovo/denovo/model.py

```diff
@@ -989,6 +990,208 @@ def configure_optimizers(
+        *args,
+    ) -> None:
+        if self.out_writer is None:
+            return
+        for (
+            indexes,
+            t_or_d,
+            peptides,
+            score_result,
+            per_aa_score,
+            precursors,
+        ) in outputs:
+            prec_mass = precursors[:, 0]
+            prec_charge = precursors[:, 1]
+            prec_mz = precursors[:, 2]
+            calc_mz = [
+                self.peptide_mass_calculator.mass(peptide, charge)
+                for peptide, charge in zip(peptides, prec_charge)
+            ]
+            for row in zip(
+                peptides,
+                score_result,
+                prec_charge,
+                prec_mz,
+                calc_mz,
+                indexes,
+                per_aa_score,
+                t_or_d,
+            ):
+                self.out_writer.psms.append(row)
+
+
+def _calc_match_score(
+    batch_all_aa_scores: torch.Tensor,
+    truth_aa_indicies: torch.Tensor,
+    decoder_reverse: bool = False,
+) -> List[float]:
+    """
+    Calculate the score between the input spectra and associated peptide.
+
+    Take in teacher-forced scoring of amino acids
+    of the peptides (in a batch) and use the truth labels
+    to calculate a score between the input spectra and
+    associated peptide. The score is the geometric
+    mean of the AA probabilities
+
+    Parameters
+    ----------
+    batch_all_aa_scores : torch.Tensor
+        Amino acid scores for all amino acids in
+        the vocabulary for every prediction made to generate
+        the associated peptide (for an entire batch)
+    truth_aa_indicies : torch.Tensor
+        Indicies of the score for each actual amino acid
+        in the peptide (for an entire batch)
+    decoder_reverse : bool
+        Whether the decoder is reversed.
+
+    Returns
+    -------
+    score : list[float], list[list[float]]
+        The score between the input spectra and associated peptide
+        (for an entire batch)
+        a list of lists of per amino acid scores
+        (for an entire batch)
+    """
+    # Remove trailing tokens from predictions based on decoder reversal
+    if decoder_reverse:
+        batch_all_aa_scores = batch_all_aa_scores[:, 1:]
+    elif not decoder_reverse:
+        batch_all_aa_scores = batch_all_aa_scores[:, :-1]
+
+    # Vectorized scoring using efficient indexing.
+    rows = (
+        torch.arange(batch_all_aa_scores.shape[0])
+        .unsqueeze(-1)
+        .expand(-1, batch_all_aa_scores.shape[1])
+    )
+    cols = torch.arange(0, batch_all_aa_scores.shape[1]).expand_as(rows)
+
+    per_aa_scores = batch_all_aa_scores[rows, cols, truth_aa_indicies]
+
+    per_aa_scores[per_aa_scores == 0] += 1e-10
+    score_mask = truth_aa_indicies != 0
+    per_aa_scores[~score_mask] = 0
+    log_per_aa_scores = torch.log(per_aa_scores)
+    all_scores = torch.where(
+        log_per_aa_scores == float("-inf"),
```

Comment:
**todo:** You don't need this `where`. You already know where the values are -inf, because that's where they were 0 before the log. And that's where you set them as such with `~score_mask`. So you can just use that mask to sum only the relevant values.